From 9cd8e85f06707168fbb982a3c3618fa119fd13b7 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Wed, 08 Dec 2021 11:46:17 +0200
Subject: [PATCH] [SW-63186] gaudi2: security: set axprot for mme

MME is configured as user level access using axprot registers.

Change-Id: I8739950405e9c91b4e34287101c392b7960178ce
---

diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index 7ed9b69..b1703be 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -44,6 +44,10 @@
 #include <dcore0_tpc0_mstr_if_core_lbw_regs.h>
 #include <dcore0_tpc6_mstr_if_core_hbw_regs.h>
 #include <dcore0_tpc6_mstr_if_core_lbw_regs.h>
+#include <dcore0_mme_sbte0_mstr_if_core_hbw_regs.h>
+#include <dcore0_mme_wb0_mstr_if_core_hbw_regs.h>
+#include <dcore0_mme_ctrl_mstr_if_core_hbw_regs.h>
+#include <dcore0_mme_ctrl_mstr_if_core_lbw_regs.h>
 
 /*
  * instances_num : Number of instances of RR to program.
@@ -208,6 +212,7 @@
 	AXPROT_ROT,
 	AXPROT_HMMU,
 	AXPROT_TPC,
+	AXPROT_MME,
 };
 
 #define DCORE_DIST	0x200000
@@ -940,6 +945,14 @@
 #define TPC_RTR_INST_DIST	0x10000
 #define TPC_INST_NUM		NUM_OF_DCORES
 #define TPC_INST_DIST		DCORE_DIST
+#define MME_SBTE_RTR_INST_NUM	5
+#define MME_SBTE_RTR_INST_DIST	0x8000
+#define MME_WB_RTR_INST_NUM	2
+#define MME_WB_RTR_INST_DIST	0x1000
+#define MME_INST_NUM		NUM_OF_DCORES
+#define MME_INST_DIST		DCORE_DIST
+#define MME_CRTL_RTR_INST_NUM	1
+
 
 static struct axprot_regs axprot_reg[] = {
 	[AXPROT_VDEC] = {
@@ -973,6 +986,14 @@
 		.dcore_num = TPC_INST_NUM,
 		.inst_dist = TPC_RTR_INST_DIST,
 		.inst_num = TPC_RTR_INST_NUM,
+	},
+	[AXPROT_MME] = {
+		.base_reg = mmDCORE0_MME_SBTE0_MSTR_IF_CORE_HBW_AXPROT_OVRD,
+		.value = AXPROT_USER_VAL,
+		.dcore_dist = MME_INST_DIST,
+		.dcore_num = MME_INST_NUM,
+		.inst_dist = MME_SBTE_RTR_INST_DIST,
+		.inst_num = MME_SBTE_RTR_INST_NUM,
 	},
 };
 
@@ -1060,6 +1081,33 @@
 	regmap_writel(AXPROT_USER_VAL, mmDCORE0_TPC6_MSTR_IF_CORE_LBW_AXPROT_OVRD);
 }
 
+static void preboot_secure_mme(void)
+{
+	/* Configure mme to user level */
+
+	/* SBTE HBW AXPROT */
+	preboot_axprot_regs_init(AXPROT_MME);
+
+	axprot_reg[AXPROT_MME].base_reg = mmDCORE0_MME_WB0_MSTR_IF_CORE_HBW_AXPROT_OVRD;
+	axprot_reg[AXPROT_MME].inst_dist = MME_WB_RTR_INST_DIST;
+	axprot_reg[AXPROT_MME].inst_num = MME_WB_RTR_INST_NUM;
+
+	/* WB HBW AXPROT */
+	preboot_axprot_regs_init(AXPROT_MME);
+
+	axprot_reg[AXPROT_MME].base_reg = mmDCORE0_MME_CTRL_MSTR_IF_CORE_HBW_AXPROT_OVRD;
+	axprot_reg[AXPROT_MME].inst_dist = 0;
+	axprot_reg[AXPROT_MME].inst_num = MME_CRTL_RTR_INST_NUM;
+
+	/* CTRL HBW AXPROT */
+	preboot_axprot_regs_init(AXPROT_MME);
+
+	axprot_reg[AXPROT_MME].base_reg = mmDCORE0_MME_CTRL_MSTR_IF_CORE_LBW_AXPROT_OVRD;
+
+	/* CTRL LBW AXPROT */
+	preboot_axprot_regs_init(AXPROT_MME);
+}
+
 void preboot_security_init(void)
 {
 	preboot_secure_pcie();
@@ -1075,6 +1123,7 @@
 		preboot_secure_arc_farm();
 		preboot_secure_trace();
 		preboot_secure_tpc();
+		preboot_secure_mme();
 		preboot_secure_lbw();
 		preboot_secure_hbw();
 		preboot_puts("Security enabled.\n");
